### PR TITLE
Async PUBACK support

### PR DESCRIPTION
**Pull request by ewertons**
Changes in umqtt to allow sending PUBACK asynchronously.

### CODE REVIEW

File: src/mqtt_client.c [hunk 1]
```diff
@@ -633,6 +633,55 @@ static int cloneMqttOptions(MQTT_CLIENT* mqtt_client, const MQTT_CLIENT_OPTIONS*
...
+        pubRel = mqtt_codec_publishAck(packetId);
+        if (pubRel == NULL)
+        {
+            LogError("Failed to allocate publish ack message.");
+            set_error_callback(mqtt_client, MQTT_CLIENT_MEMORY_ERROR);
+        }
+
+        response_packet_type = PUBACK_TYPE;
+    }
+    if (pubRel != NULL)
```
Comment (danewalton) [inline src/mqtt_client.c:662]: nit: newline above this to break up the if/then blocks

File: src/mqtt_client.c [hunk 2]
```diff
@@ -633,6 +633,55 @@ static int cloneMqttOptions(MQTT_CLIENT* mqtt_client, const MQTT_CLIENT_OPTIONS*
     return result;
 }
 
+static void SendMessageAck(MQTT_CLIENT* mqtt_client, uint16_t packetId, QOS_VALUE qosValue)
+{
+    CONTROL_PACKET_TYPE response_packet_type = UNKNOWN_TYPE;
+    BUFFER_HANDLE pubRel = NULL;
+    if (qosValue == DELIVER_EXACTLY_ONCE)
+    {
+        pubRel = mqtt_codec_publishReceived(packetId);
```
Comment (danewalton) [inline src/mqtt_client.c:642]: I'm a little confused here. Why is a **Send**MessageAck doing a `rec` here?

Comment (ewertons) [reply src/mqtt_client.c:642]: I did not want the upper layer to have to decide between ACK and REC, so I created one function that takes the QoS. I agree the name might not be clear, we could change that.

Comment (danewalton) [reply src/mqtt_client.c:642]: maybe rename to `SendMessageResponse`?

Comment (ewertons) [reply src/mqtt_client.c:642]: I think that's better indeed. "Response" is the best term as it is exactly the one mentioned by the mqtt spec:

http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/csprd02/mqtt-v3.1.1-csprd02.html#_Toc385349777



<h5 style="margin: 12pt 0in 6pt; text-indent: 0in; break-after: avoid; font-size: 12pt; font-family: Arial, sans-serif; color: rgb(59, 0, 111); font-weight: bold; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><span style="font-size: 10pt;">Table 3.4 - Expected Publish Packet response</span></h5>
QoS Level | Expected Response
-- | --
QoS 0 | None
QoS 1 | PUBACK Packet
QoS 2 | PUBREC Packet

<p class="MsoNormal" style="margin: 4pt 0in; font-size: 10pt; font-family: Arial, sans-serif; color: rgb(0, 0, 0); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">&nbsp;</p>

Comment (ewertons) [reply src/mqtt_client.c:642]: The final function became: `mqtt_client_send_message_response`

**Review by danewalton — APPROVED** [review]

**Pull request merged**